examples/vector_database/scripts/compute_vectors.py

```diff
@@ -0,0 +1,240 @@
+        self.pretrained = pretrained
+        self.device = device or ('cuda' if torch.cuda.is_available() else 'cpu')
+        self.dataset_name = dataset_name
+        self.split = split
+        self.streaming = streaming
+        self.model = None
+        self.preprocess = None
+
+    async def setup_model(self):
+        """Set up the CLIP model."""
+        import open_clip
+
+        model, _, preprocess = open_clip.create_model_and_transforms(
+            self.model_name, pretrained=self.pretrained, device=self.device)
+        self.model = model
+        self.preprocess = preprocess
+
+    async def get_dataset_iterator(self) -> AsyncIterator[Tuple[int, Any]]:
+        """Load data from a HuggingFace dataset."""
+        from datasets import load_dataset
+
+        dataset = load_dataset(self.dataset_name,
+                               streaming=self.streaming,
+                               trust_remote_code=True)[self.split]
+
+        if self.start_idx > 0:
+            dataset = dataset.skip(self.start_idx)
+
+        for idx, item in enumerate(dataset, start=self.start_idx):
+            if self.end_idx and idx >= self.end_idx:
+                break
+            yield idx, item
+
+    async def do_data_loading(
+            self) -> AsyncIterator[Tuple[int, Tuple[torch.Tensor, Any]]]:
+        """Load and preprocess ImageNet images."""
+        if self.model is None:
+            await self.setup_model()
+
+        async for idx, item in self.get_dataset_iterator():
+            try:
+                # ImageNet provides PIL Images directly
+                tensor = self.preprocess(item['image'])
+                if tensor is not None:
+                    # Pass through both the tensor and original image
+                    yield idx, (tensor, item['image'])
+            except Exception as e:
+                logging.debug(
+                    f"Error preprocessing image at index {idx}: {str(e)}")
+
+    async def do_batch_processing(
+        self, batch: List[Tuple[int, Tuple[torch.Tensor, Any]]]
+    ) -> List[Tuple[int, bytes]]:
+        """Process a batch of images through CLIP."""
+        if self.model is None:
+            await self.setup_model()
+
+        # Unpack the batch
+        indices, batch_data = zip(*batch)
+        model_inputs, original_images = zip(*batch_data)
+
+        # Stack inputs into a batch
+        batch_tensor = torch.stack(model_inputs).to(self.device)
+
+        # Run inference
+        with torch.no_grad():
+            features = self.model.encode_image(batch_tensor)
+            features /= features.norm(dim=-1, keepdim=True)
+
+        # Convert to numpy arrays
+        embeddings = features.cpu().numpy()
+
+        # Convert original images to base64
+        images_base64 = {}
+        for idx, img in zip(indices, original_images):
+            buffered = BytesIO()
+            img.save(buffered, format="JPEG")
+            img_str = base64.b64encode(buffered.getvalue()).decode()
```

Comment:
can we not just use an unencoded bytestring here? it doesn't hugely matter but if we can simplify the code, we should